From 43185fc46573109dff12e046fbac91c5c45f6d96 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:31 -0400
Subject: [fs] nfs: add checks for returned value of try_module_get()

Message-id: <1411050057-12417-69-git-send-email-steved@redhat.com>
Patchwork-id: 96208
O-Subject: [RHEL7.1 PATCH 68/94] NFS: add checks for returned value of try_module_get()
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Alexey Khoroshilov <khoroshilov@ispras.ru>

There is a couple of places in client code where returned value
of try_module_get() is ignored. As a result there is a small chance
to premature unload module because of unbalanced refcounting.

The patch adds error handling in that places.

Found by Linux Driver Verification project (linuxtesting.org).

Signed-off-by: Alexey Khoroshilov <khoroshilov@ispras.ru>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 1f70ef96b176bdb3b75230ec68850d83736b387b
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/client.c b/fs/nfs/client.c
index 168aa0d..b7bfa27 100644
--- a/fs/nfs/client.c
+++ b/fs/nfs/client.c
@@ -110,8 +110,8 @@ struct nfs_subversion *get_nfs_version(unsigned int version)
   mutex_unlock(&nfs_version_mutex);
  }
 
- if (!IS_ERR(nfs))
-  try_module_get(nfs->owner);
+ if (!IS_ERR(nfs) && !try_module_get(nfs->owner))
+  return ERR_PTR(-EAGAIN);
  return nfs;
 }
 
@@ -158,7 +158,8 @@ struct nfs_client *nfs_alloc_client(const struct nfs_client_initdata *cl_init)
   goto error_0;
 
  clp->cl_nfs_mod = cl_init->nfs_mod;
- try_module_get(clp->cl_nfs_mod->owner);
+ if (!try_module_get(clp->cl_nfs_mod->owner))
+  goto error_dealloc;
 
  clp->rpc_ops = clp->cl_nfs_mod->rpc_ops;
 
@@ -190,6 +191,7 @@ struct nfs_client *nfs_alloc_client(const struct nfs_client_initdata *cl_init)
 
 error_cleanup:
  put_nfs_version(clp->cl_nfs_mod);
+error_dealloc:
  kfree(clp);
 error_0:
  return ERR_PTR(err);
-- 
1.7.1